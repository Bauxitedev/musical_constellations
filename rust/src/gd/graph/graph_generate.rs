use std::{any::type_name, collections::BTreeMap, f64::consts::TAU, num::NonZero};

use godot::prelude::*;
use kiddo::{NearestNeighbour, SquaredEuclidean, float::kdtree::KdTree};
use nalgebra::{Point3, Unit, UnitQuaternion, Vector3 as NVector3};
use ordered_float::OrderedFloat;
use petgraph::{
    Graph,
    algo::tarjan_scc,
    graph::{NodeIndex, UnGraph},
};
use rand::{Rng, SeedableRng as _, seq::IndexedRandom as _};
use rand_xoshiro::Xoshiro256Plus;
use serde::{Deserialize, Serialize};
use strum::IntoEnumIterator as _;
use tracing::instrument;

use crate::{chords::Chord, gd::graph::graph_main::GraphTypedef, profile, util::random_unit_axis};

#[derive(Debug, Serialize, Deserialize)]
pub struct ConstellationGraph {
    pub chord: Chord,
    pub semitone_offset: i32,
    pub graph: GraphTypedef,
    pub islands: Vec<Vec<NodeIndex>>, // Strongly Connected Components (aka "islands")
}

impl ConstellationGraph {
    /// Create the graph and its strongly connected components (islands)
    pub fn new<R: Rng>(n: usize, radius: f32, max_neighbor_count: usize, rng: &mut R) -> Self {
        tracing::info!(rng_type = type_name::<R>(), "generating ConstellationGraph");

        // Generate this first for rng reasons
        let chords = Chord::iter().collect::<Vec<_>>();
        let chord = *chords
            .choose(&mut Xoshiro256Plus::from_rng(rng)) // Making a new rng here to avoid nondeterminism when we change the amount of chords
            .unwrap();
        let semitone_offset_base = rng.random_range(-11..12); // Equal for all notes to avoid dissonance

        let points = Self::generate_points(n, radius as f64, rng);

        let voronoi_rng = Xoshiro256Plus::from_rng(rng);
        let clusters = profile!(
            "cluster_voronoi",
            Self::cluster_voronoi(
                points,
                (n as f64 / 15.0).ceil() as usize, // n / 15 means ~15 nodes per cluster
                voronoi_rng
            )
        );

        let supergraph = Self::connect_clusters_internally(&clusters, max_neighbor_count, rng);
        let scc = tarjan_scc(&supergraph);

        ConstellationGraph {
            chord,
            semitone_offset: semitone_offset_base,
            graph: supergraph,
            islands: scc,
        }
    }

    /// Generate n random points on the unit sphere
    fn generate_points<R: Rng>(n: usize, radius: f64, rng: &mut R) -> Vec<Vector3> {
        // Using a large value of `max_angle` here is safe now due to our new `leniency` algorithm
        let max_angle = TAU * 0.02;
        let points_nalgebra = Self::generate_points_poisson(n, radius, max_angle, rng);

        points_nalgebra
            .into_iter()
            .map(|p| Vector3::from_array(<[f32; 3]>::from(p.cast::<f32>())))
            .collect()
    }

    /// Generates `n` points on the surface of a sphere of radius `radius` using a Poisson-disk-like algorithm.
    /// The points are roughly `max_angle` radians separated from each other, unless the algorithm reaches an iteration limit.
    /// Then it will slowly reduce the angle limit until it succeeds. This is called `leniency`.
    #[cfg_attr(feature = "enable-tracing", instrument(skip(rng)))]
    fn generate_points_poisson<R: Rng>(
        n: usize,
        radius: f64,
        mut max_angle: f64,
        rng: &mut R,
    ) -> Vec<Point3<f64>> {
        pub type KdTreeUsize<A, const K: usize> = KdTree<A, usize, K, 32, u32>;

        let mut points = Vec::with_capacity(n);
        let mut tree = KdTreeUsize::new();

        let epsilon_mult = 0.9; // This is the "leniency factor"

        // Starting point: (0, radius, 0)
        let start = NVector3::new(0.0, radius, 0.0);
        let start_point = Point3::from(start);
        tree.add(&start_point.into(), 0);
        points.push(start_point);

        while points.len() < n {
            let new_point;

            // Random angle per point at 50%..150% of max_angle (this is important to ensure melody rest times are varied - they depend on inter-node distance)
            let angle = rng.random_range(0.5..=1.5) * max_angle;

            // We want chord distance formula here, NOT arc length (we want an underestimation, not an overestimation)
            let mut min_distance = 2.0 * radius * (angle / 2.0).sin() * epsilon_mult;

            let mut iteration = 0;
            loop {
                // Select a previous point as a base
                let parent_idx = rng.random_range(0..points.len());
                let parent = points[parent_idx].coords;

                // If iteration count goes beyond 1k or so, reduce minimum distance (becomes slightly more `lenient`)
                if iteration >= 1_000 {
                    iteration = 0;

                    let leniency_factor = 0.9;
                    min_distance *= leniency_factor;
                    max_angle *= leniency_factor;

                    tracing::warn!(
                        "Reducing poisson min_distance to {:.4} due to reaching iteration limit",
                        min_distance
                    );
                }

                // Apply random axis rotation by fixed angle
                let axis = random_unit_axis(rng);
                let rot = UnitQuaternion::from_axis_angle(&Unit::new_normalize(axis), angle);
                let candidate = rot * parent;

                // Find closest neighbor
                let neighbors = tree.nearest_n_within::<SquaredEuclidean>(
                    &candidate.into(),
                    min_distance * min_distance,
                    NonZero::new(1).unwrap(), //Return up to 1 result
                    false,
                );

                // Reject if too close to neighbor
                if neighbors.is_empty() {
                    let p = Point3::from(candidate);
                    tree.add(&p.into(), points.len());
                    new_point = p;
                    break;
                }

                iteration += 1;
            }

            points.push(new_point);
        }

        points
    }

    /// Cluster the points according to a voronoi-like algorithm, where `k` is the amount of clusters. Returns the clustered points and the cluster centroids.
    #[cfg_attr(feature = "enable-tracing", instrument(skip(points, rng)))]
    fn cluster_voronoi(
        points: Vec<Vector3>,
        k: usize,
        mut rng: Xoshiro256Plus,
    ) -> Vec<(Vec<Vector3>, Vector3)> {
        pub type KdTreeUsize<A, const K: usize> = KdTree<A, usize, K, 32, u32>;
        let mut kdtree = KdTreeUsize::new();

        // Pick k random clusters
        let centroids = points.choose_multiple(&mut rng, k).collect::<Vec<_>>();

        // Add them to kd-tree
        for (i, c) in centroids.iter().enumerate() {
            kdtree.add(&[c.x, c.y, c.z], i);
        }

        let mut clusters = centroids
            .iter()
            .map(|centroid| (vec![], **centroid))
            .collect::<Vec<_>>();

        // Find closest centroid per point and assign the point to that cluster
        for p in points {
            let closest_centroid = kdtree
                .nearest_one::<SquaredEuclidean>(&[p.x, p.y, p.z])
                .item;
            clusters[closest_centroid].0.push(p);
        }

        //Now sort by centroid y (stable sort) for cool animation!
        clusters.sort_by_key(|cluster| OrderedFloat(-cluster.1.y));

        clusters
    }

    fn connect_clusters_internally<R: Rng>(
        clusters: &[(Vec<Vector3>, Vector3)],
        max_neighbor_count: usize,
        rng: &mut R,
    ) -> GraphTypedef {
        pub type KdTreeUsize<A, const K: usize> = KdTree<A, usize, K, 32, u32>;

        let mut supergraph: GraphTypedef = Graph::new_undirected();
        let mut graphs = vec![];

        profile!(
            "connect_clusters_internally", // This is not the bottleneck btw
            for (cluster, _centroid) in clusters.iter() {
                // Build the kd-tree for nearest neighbor search per-cluster
                let mut kdtree = KdTreeUsize::new();
                for (i, p) in cluster.iter().enumerate() {
                    kdtree.add(&[p.x, p.y, p.z], i);
                }

                // Create a graph where each point is a node
                let mut graph: GraphTypedef = Graph::new_undirected();
                let node_indices: Vec<NodeIndex> =
                    cluster.iter().map(|p| graph.add_node(*p)).collect();

                // Add edges between each node and its 1..max_neighbor_count nearest neighbors
                let max_neighbor_count = max_neighbor_count as i32;
                let min_neighbor_count = 1;

                for (i, point) in cluster.iter().enumerate() {
                    let neighbor_count = *(min_neighbor_count..=max_neighbor_count)
                        .collect::<Vec<_>>()
                        .choose_weighted(rng, |w| match w {
                            1 => 0.8,
                            2 => 1.0,
                            3 => 0.3,
                            _ => 0.0,
                        })
                        .unwrap() as usize;

                    // Find neighbors
                    let query = [point.x, point.y, point.z];
                    let neighbors =
                        kdtree.nearest_n::<SquaredEuclidean>(&query, neighbor_count + 1); //+1 since we get the point itself too

                    // Connect vertex i to neighbor j
                    for NearestNeighbour { item: j, .. } in &neighbors {
                        if i != *j {
                            let a = node_indices[i];
                            let b = node_indices[*j];

                            //Only add edge if there isn't one already
                            if !graph.contains_edge(a, b) {
                                graph.add_edge(a, b, ());
                            }
                        }
                    }
                }

                graphs.push(graph);
            }
        );

        Self::merge_undirected_graphs(&mut supergraph, &graphs);
        supergraph
    }

    /// Merges multiple undirected graphs together into one big graph.
    fn merge_undirected_graphs<N: Clone, E: Clone>(
        base: &mut UnGraph<N, E>,
        others: &[UnGraph<N, E>],
    ) {
        for g in others {
            //Use BTreeMap for determinism
            let mut node_map: BTreeMap<NodeIndex, NodeIndex> = BTreeMap::new();

            // Add all nodes from current graph 'g' into 'base'
            for node in g.node_indices() {
                let new_node = base.add_node(g[node].clone());
                node_map.insert(node, new_node);
            }

            // Add all edges, updating node indices to new ones in 'base'
            for edge in g.edge_indices() {
                let (source, target) = g.edge_endpoints(edge).unwrap();
                let weight = g.edge_weight(edge).unwrap().clone();
                base.add_edge(node_map[&source], node_map[&target], weight);
            }
        }
    }
}

impl PartialEq for ConstellationGraph {
    fn eq(&self, other: &Self) -> bool {
        graph_eq(&self.graph, &other.graph) && self.islands == other.islands
    }
}

/// Check if two graphs are "equal". Checks if internal node/edges have the exact same order. Useful for determinism testing.
/// See https://github.com/petgraph/petgraph/issues/199#issuecomment-484077775
fn graph_eq<N, E, Ty, Ix>(
    a: &petgraph::Graph<N, E, Ty, Ix>,
    b: &petgraph::Graph<N, E, Ty, Ix>,
) -> bool
where
    N: PartialEq,
    E: PartialEq,
    Ty: petgraph::EdgeType,
    Ix: petgraph::graph::IndexType + PartialEq,
{
    let a_ns = a.raw_nodes().iter().map(|n| &n.weight);
    let b_ns = b.raw_nodes().iter().map(|n| &n.weight);
    let a_es = a
        .raw_edges()
        .iter()
        .map(|e| (e.source(), e.target(), &e.weight));
    let b_es = b
        .raw_edges()
        .iter()
        .map(|e| (e.source(), e.target(), &e.weight));
    a_ns.eq(b_ns) && a_es.eq(b_es)
}
