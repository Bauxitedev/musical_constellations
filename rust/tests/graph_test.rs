//! This module tests determinism of the graph constellation generation algorithm.
//! Always use `Xoshiro256Plus` to guarantee portability/determinism across all platforms.
//! Also don't use the built-in hash `DefaultHash` or `ahash`, try `HighwayHash` instead (it's fully portable/deterministic).
//! Also watch out for HashMap/HashSet, by default they're randomized.

use musical_constellations_rust::gd::graph::graph_generate::ConstellationGraph;
use rand::Rng;
use serde::Serialize;
#[derive(Serialize)]
pub struct ConstellationGraphSnapshot {
    global_seed: i64,
    num_points: usize,
    max_neighbor_count: usize,
    radius: f32,
    rng_type: String,

    constellation_graph: ConstellationGraph,
}

impl ConstellationGraphSnapshot {
    pub fn new<R: Rng>(
        constellation_graph: ConstellationGraph,
        global_seed: i64,
        radius: f32,
        max_neighbor_count: usize,
        _rng: R,
    ) -> Self {
        Self {
            global_seed,
            num_points: constellation_graph.graph.node_count(),
            constellation_graph,
            radius,
            max_neighbor_count,
            rng_type: std::any::type_name::<R>().to_owned(),
        }
    }
}

// Unit tests
#[cfg(test)]
mod tests {
    use rand::SeedableRng as _;
    use rand_xoshiro::Xoshiro256Plus;

    use super::*; // bring functions from outer scope

    /// This generates a ConstellationGraph with fixed seeds and compares them to the snapshot.
    /// Run this test on many different platforms/targets/profiles, to ensure the graph generation process doesn't suffer from float nondeterminism.
    #[test]
    fn graph_determinism() {
        let mut snapshots = vec![];

        for global_seed in [1_i64, 2, i64::MAX, i64::MIN] {
            let num_points = 30; // Do not use 2000 here, the files become too unwieldy
            let max_neighbor_count = 2 - 1;
            let radius = 5.0;

            let mut seed_bytes = [0u8; 32];
            seed_bytes[0..8].copy_from_slice(&global_seed.to_le_bytes());

            let mut rng = Xoshiro256Plus::from_seed(seed_bytes);
            let constellation_graph =
                ConstellationGraph::new(num_points as usize, radius, max_neighbor_count, &mut rng);
            let snapshot = ConstellationGraphSnapshot::new(
                constellation_graph,
                global_seed,
                radius,
                max_neighbor_count,
                rng,
            );

            snapshots.push(snapshot);
        }

        insta::assert_yaml_snapshot!(snapshots);
    }
}
