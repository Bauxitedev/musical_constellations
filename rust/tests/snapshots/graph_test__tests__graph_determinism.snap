---
source: tests/graph_test.rs
assertion_line: 71
expression: snapshots
---
- global_seed: 1
  num_points: 30
  max_neighbor_count: 1
  radius: 5
  rng_type: "rand_xoshiro::xoshiro256plus::Xoshiro256Plus"
  constellation_graph:
    chord: Cmaj
    semitone_offset: -6
    graph:
      nodes:
        - x: 0
          y: 5
          z: 0
        - x: -0.48186162
          y: 4.966032
          z: 0.3260921
        - x: -0.8538491
          y: 4.8905325
          z: 0.5946726
        - x: -0.35522574
          y: 4.987238
          z: -0.03564722
        - x: -0.3267249
          y: 4.9122515
          z: 0.87352014
        - x: -0.5560027
          y: 4.9505806
          z: -0.4273327
        - x: 0.3978722
          y: 4.9512205
          z: 0.57193977
        - x: -1.2500768
          y: 4.839121
          z: 0.14218372
        - x: 1.0611526
          y: 4.801889
          z: 0.9032267
        - x: -0.45250168
          y: 4.8056407
          z: 1.304247
        - x: 0.09877963
          y: 4.932452
          z: 0.8131157
        - x: 0.2315436
          y: 4.865633
          z: 1.1278311
        - x: -1.1213691
          y: 4.603559
          z: 1.596801
        - x: -0.068105005
          y: 4.687251
          z: 1.7392638
        - x: -1.9199497
          y: 4.59352
          z: -0.46191677
        - x: -1.9760897
          y: 4.420082
          z: -1.2481761
        - x: -1.7792746
          y: 4.2693152
          z: -1.8992445
        - x: -2.5851376
          y: 4.218601
          z: -0.72143483
        - x: -2.4742107
          y: 4.16039
          z: -1.2527717
        - x: -0.91771626
          y: 4.9138036
          z: -0.11105234
        - x: -1.5521612
          y: 4.6921887
          z: 0.7577346
        - x: -2.1830754
          y: 4.4980025
          z: 0.046441637
        - x: -2.4439163
          y: 4.3532896
          z: 0.27593747
        - x: 1.4552125
          y: 4.7598276
          z: 0.47581244
        - x: -1.6163744
          y: 4.549238
          z: 1.3006785
        - x: -0.5708639
          y: 4.419575
          z: 2.2674813
        - x: -0.15756111
          y: 4.201559
          z: 2.7059333
        - x: -0.098641604
          y: 3.744413
          z: 3.3120449
        - x: -1.2852378
          y: 4.270076
          z: 2.261552
        - x: -0.18478894
          y: 4.467515
          z: 2.2376692
      node_holes: []
      edge_property: undirected
      edges:
        - - 0
          - 3
          - ~
        - - 1
          - 3
          - ~
        - - 2
          - 1
          - ~
        - - 4
          - 10
          - ~
        - - 5
          - 3
          - ~
        - - 6
          - 10
          - ~
        - - 7
          - 19
          - ~
        - - 8
          - 23
          - ~
        - - 9
          - 4
          - ~
        - - 10
          - 11
          - ~
        - - 12
          - 24
          - ~
        - - 13
          - 9
          - ~
        - - 14
          - 21
          - ~
        - - 15
          - 18
          - ~
        - - 16
          - 15
          - ~
        - - 17
          - 18
          - ~
        - - 20
          - 24
          - ~
        - - 21
          - 22
          - ~
        - - 25
          - 29
          - ~
        - - 26
          - 29
          - ~
        - - 27
          - 26
          - ~
        - - 28
          - 25
          - ~
    islands:
      - - 5
        - 2
        - 1
        - 3
        - 0
      - - 11
        - 6
        - 10
        - 13
        - 9
        - 4
      - - 19
        - 7
      - - 23
        - 8
      - - 20
        - 24
        - 12
      - - 22
        - 21
        - 14
      - - 17
        - 18
        - 16
        - 15
      - - 27
        - 26
        - 29
        - 28
        - 25
- global_seed: 2
  num_points: 30
  max_neighbor_count: 1
  radius: 5
  rng_type: "rand_xoshiro::xoshiro256plus::Xoshiro256Plus"
  constellation_graph:
    chord: Cmaj
    semitone_offset: 0
    graph:
      nodes:
        - x: 0
          y: 5
          z: 0
        - x: 0.51347935
          y: 4.9592247
          z: 0.37740046
        - x: -0.95646846
          y: 4.787545
          z: -1.0791564
        - x: -0.08573956
          y: 4.9276657
          z: -0.8430668
        - x: 0.3221012
          y: 4.9895434
          z: -0.026592748
        - x: 0.05578106
          y: 4.967373
          z: -0.56753236
        - x: -0.11918026
          y: 4.990678
          z: 0.28094438
        - x: -0.6554796
          y: 4.950409
          z: 0.25258636
        - x: -1.2920094
          y: 4.796442
          z: 0.5699624
        - x: -0.80451787
          y: 4.8598647
          z: 0.85700905
        - x: -0.5675702
          y: 4.8507705
          z: 1.0713956
        - x: -1.1511054
          y: 4.851611
          z: -0.36989847
        - x: 0.2043493
          y: 4.9424953
          z: 0.72799784
        - x: -0.48370263
          y: 4.578451
          z: 1.9503381
        - x: -1.1042429
          y: 4.4088197
          z: 2.0839763
        - x: -1.4189234
          y: 4.754163
          z: -0.620154
        - x: -0.0046411017
          y: 4.8401113
          z: 1.2543135
        - x: -1.6616176
          y: 4.5338078
          z: -1.2975422
        - x: -2.1606033
          y: 4.4619746
          z: -0.6500577
        - x: -1.5035919
          y: 4.433586
          z: 1.7557126
        - x: -2.7831063
          y: 4.1283917
          z: -0.45902228
        - x: -2.1559083
          y: 4.33504
          z: -1.2487944
        - x: -2.4989133
          y: 4.321078
          z: 0.2893417
        - x: -0.39911416
          y: 4.9589467
          z: 0.49955752
        - x: -1.908052
          y: 4.6061993
          z: 0.3771818
        - x: -2.8210235
          y: 3.8292418
          z: -1.5423145
        - x: 0.64073896
          y: 4.9082775
          z: 0.7058776
        - x: -2.3170142
          y: 4.417363
          z: -0.3440193
        - x: -2.3786879
          y: 3.9417129
          z: -1.9505752
        - x: -2.193154
          y: 4.082068
          z: 1.8779763
      node_holes: []
      edge_property: undirected
      edges:
        - - 0
          - 4
          - ~
        - - 1
          - 4
          - ~
        - - 2
          - 3
          - ~
        - - 3
          - 5
          - ~
        - - 6
          - 23
          - ~
        - - 7
          - 23
          - ~
        - - 8
          - 9
          - ~
        - - 9
          - 10
          - ~
        - - 11
          - 15
          - ~
        - - 12
          - 26
          - ~
        - - 13
          - 14
          - ~
        - - 14
          - 19
          - ~
        - - 16
          - 12
          - ~
        - - 17
          - 21
          - ~
        - - 18
          - 27
          - ~
        - - 20
          - 27
          - ~
        - - 22
          - 24
          - ~
        - - 25
          - 28
          - ~
        - - 29
          - 19
          - ~
    islands:
      - - 1
        - 4
        - 0
      - - 5
        - 3
        - 2
      - - 7
        - 23
        - 6
      - - 10
        - 9
        - 8
      - - 15
        - 11
      - - 26
        - 16
        - 12
      - - 29
        - 19
        - 14
        - 13
      - - 21
        - 17
      - - 20
        - 27
        - 18
      - - 24
        - 22
      - - 28
        - 25
- global_seed: 9223372036854775807
  num_points: 30
  max_neighbor_count: 1
  radius: 5
  rng_type: "rand_xoshiro::xoshiro256plus::Xoshiro256Plus"
  constellation_graph:
    chord: Cmaj
    semitone_offset: -6
    graph:
      nodes:
        - x: 0
          y: 5
          z: 0
        - x: 0.1412346
          y: 4.961727
          z: -0.60109556
        - x: 0.5348556
          y: 4.8641787
          z: -1.0264969
        - x: -0.07267101
          y: 4.8257055
          z: -1.3066304
        - x: -0.1853845
          y: 4.945359
          z: -0.7134805
        - x: 0.4825057
          y: 4.9665294
          z: 0.31745267
        - x: -0.9008619
          y: 4.9160266
          z: 0.14536564
        - x: -0.9811117
          y: 4.852913
          z: -0.69760895
        - x: 0.81664115
          y: 4.5860834
          z: -1.8168484
        - x: -0.7500314
          y: 4.6703024
          z: -1.6204095
        - x: -1.3981055
          y: 4.7834463
          z: 0.40489542
        - x: 0.71368736
          y: 4.3723783
          z: -2.3179638
        - x: -1.9707779
          y: 4.587135
          z: -0.27244595
        - x: -1.5994989
          y: 4.57935
          z: 1.2129124
        - x: -0.35877237
          y: 4.4785123
          z: -2.1941311
        - x: -2.597455
          y: 4.272096
          z: -0.049226496
        - x: -1.2219007
          y: 4.6621246
          z: -1.3309968
        - x: 0.2771477
          y: 4.1706886
          z: -2.7438192
        - x: -2.1859913
          y: 4.331632
          z: 1.2076448
        - x: -0.0949389
          y: 3.724092
          z: -3.334985
        - x: -2.2019346
          y: 4.465036
          z: -0.46361357
        - x: 1.2598548
          y: 4.656683
          z: -1.31456
        - x: 2.0765336
          y: 4.421378
          z: -1.0674374
        - x: 1.5299664
          y: 4.7341757
          z: -0.49677458
        - x: 2.0183976
          y: 4.1661477
          z: -1.8892545
        - x: 1.2218667
          y: 4.8358173
          z: 0.34916067
        - x: 2.1236212
          y: 4.5090494
          z: -0.39838094
        - x: 2.2753925
          y: 3.8177133
          z: -2.290776
        - x: 2.6423368
          y: 4.1345034
          z: -0.9612159
        - x: 2.9496682
          y: 3.75152
          z: -1.4918294
      node_holes: []
      edge_property: undirected
      edges:
        - - 0
          - 5
          - ~
        - - 1
          - 4
          - ~
        - - 2
          - 1
          - ~
        - - 3
          - 4
          - ~
        - - 6
          - 10
          - ~
        - - 7
          - 16
          - ~
        - - 8
          - 11
          - ~
        - - 9
          - 16
          - ~
        - - 12
          - 20
          - ~
        - - 13
          - 18
          - ~
        - - 14
          - 9
          - ~
        - - 15
          - 20
          - ~
        - - 17
          - 11
          - ~
        - - 19
          - 17
          - ~
        - - 21
          - 23
          - ~
        - - 22
          - 28
          - ~
        - - 23
          - 26
          - ~
        - - 24
          - 27
          - ~
        - - 25
          - 23
          - ~
        - - 29
          - 28
          - ~
    islands:
      - - 5
        - 0
      - - 3
        - 4
        - 2
        - 1
      - - 10
        - 6
      - - 14
        - 9
        - 16
        - 7
      - - 19
        - 17
        - 11
        - 8
      - - 15
        - 20
        - 12
      - - 18
        - 13
      - - 26
        - 25
        - 23
        - 21
      - - 29
        - 28
        - 22
      - - 27
        - 24
- global_seed: -9223372036854775808
  num_points: 30
  max_neighbor_count: 1
  radius: 5
  rng_type: "rand_xoshiro::xoshiro256plus::Xoshiro256Plus"
  constellation_graph:
    chord: Cmaj
    semitone_offset: -11
    graph:
      nodes:
        - x: 0
          y: 5
          z: 0
        - x: -0.21598676
          y: 4.99038
          z: 0.22239505
        - x: -0.18855023
          y: 4.9408555
          z: -0.7432324
        - x: -0.9515528
          y: 4.908412
          z: 0.045169752
        - x: -1.2335254
          y: 4.7892714
          z: -0.73572904
        - x: -1.0324044
          y: 4.847026
          z: 0.663687
        - x: -0.019000944
          y: 4.8925695
          z: -1.0307301
        - x: -0.38784206
          y: 4.9218154
          z: 0.79076767
        - x: -0.34659067
          y: 4.8395925
          z: 1.2075673
        - x: -1.5730757
          y: 4.591612
          z: -1.201056
        - x: -1.7725805
          y: 4.5885005
          z: 0.89644945
        - x: -1.2928675
          y: 4.8196836
          z: -0.31486896
        - x: 0.124353416
          y: 4.9528365
          z: 0.6737563
        - x: -0.49386477
          y: 4.7929106
          z: -1.3357047
        - x: -1.4594023
          y: 4.700639
          z: 0.8798529
        - x: -1.025126
          y: 4.622132
          z: 1.6077983
        - x: -1.7389216
          y: 4.2566447
          z: -1.9639573
        - x: -0.19390377
          y: 4.7567096
          z: 1.5284351
        - x: -1.6154941
          y: 4.271234
          z: 2.036354
        - x: 0.39179498
          y: 4.9755354
          z: -0.30090803
        - x: -1.6079359
          y: 4.696725
          z: 0.59608686
        - x: -0.5250425
          y: 4.9640336
          z: -0.28758094
        - x: -2.0569863
          y: 4.2722898
          z: 1.5863001
        - x: 0.416039
          y: 4.67054
          z: -1.73579
        - x: 1.0851308
          y: 4.3113422
          z: -2.2879722
        - x: 0.103920735
          y: 4.57941
          z: -2.0045462
        - x: -0.2721233
          y: 4.2214813
          z: -2.6655293
        - x: 0.15181817
          y: 4.3592095
          z: -2.444227
        - x: -0.073976174
          y: 4.462075
          z: -2.2548642
        - x: -1.1788684
          y: 4.20758
          z: -2.4303374
      node_holes: []
      edge_property: undirected
      edges:
        - - 0
          - 1
          - ~
        - - 2
          - 6
          - ~
        - - 3
          - 11
          - ~
        - - 4
          - 11
          - ~
        - - 5
          - 14
          - ~
        - - 7
          - 8
          - ~
        - - 8
          - 17
          - ~
        - - 9
          - 4
          - ~
        - - 10
          - 14
          - ~
        - - 12
          - 7
          - ~
        - - 13
          - 6
          - ~
        - - 14
          - 20
          - ~
        - - 15
          - 18
          - ~
        - - 16
          - 9
          - ~
        - - 18
          - 22
          - ~
        - - 19
          - 0
          - ~
        - - 21
          - 3
          - ~
        - - 23
          - 25
          - ~
        - - 24
          - 23
          - ~
        - - 25
          - 28
          - ~
        - - 26
          - 27
          - ~
        - - 27
          - 28
          - ~
        - - 29
          - 26
          - ~
    islands:
      - - 1
        - 19
        - 0
      - - 13
        - 6
        - 2
      - - 16
        - 9
        - 4
        - 11
        - 21
        - 3
      - - 20
        - 10
        - 14
        - 5
      - - 17
        - 8
        - 12
        - 7
      - - 22
        - 18
        - 15
      - - 29
        - 26
        - 27
        - 28
        - 25
        - 24
        - 23
